contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,582 @@
+     */
+    function _sellComponentsForOutputToken(ISetToken _setToken, uint256 _amountSetToken, ZeroExSwapQuote[] memory _swaps, IERC20 _outputToken, address _issuanceModule)
+        internal
+        returns (uint256 totalOutputTokenBought)
+    {
+        (address[] memory components, uint256[] memory componentUnits) = getRequiredRedemptionComponents(_issuanceModule, _setToken, _amountSetToken);
+        for (uint256 i = 0; i < _swaps.length; i++) {
+            require(components[i] == address(_swaps[i].sellToken), "ExchangeIssuance: COMPONENT / QUOTE ADDRESS MISMATCH");
+            require(address(_swaps[i].buyToken) == address(_outputToken), "ExchangeIssuance: INVALID BUY TOKEN");
+            uint256 maxAmountSell = componentUnits[i];
+
+            require(maxAmountSell <= IERC20(components[i]).balanceOf(address(this)), "ExchangeIssuance: INSUFFICIENT COMPONENT BALANCE");
+
+            uint256 outputTokenAmountBought;
+            uint256 componentAmountSold;
+
+            // If the component is equal to the input token we don't have to trade
+            if(components[i] == address(_swaps[i].buyToken)) {
+                outputTokenAmountBought = maxAmountSell;
+                componentAmountSold = maxAmountSell;
+            }
+            else {
+                _safeApprove(_swaps[i].sellToken, address(swapTarget), maxAmountSell);
+                (outputTokenAmountBought, componentAmountSold) = _fillQuote(_swaps[i]);
+            }
+
+            require(maxAmountSell >= componentAmountSold, "ExchangeIssuance: OVERSOLD COMPONENT");
+            totalOutputTokenBought = totalOutputTokenBought.add(outputTokenAmountBought);
+        }
+    }
+
+    /**
+     * Execute a 0x Swap quote
+     *
+     * @param _quote          Swap quote as returned by 0x API
+     *
+     * @return boughtAmount   The amount of _quote.buyToken obtained
+     * @return spentAmount    The amount of _quote.sellToken spent
+     */
+    function _fillQuote(
+        ZeroExSwapQuote memory _quote
+    )
+        internal
+        returns(uint256 boughtAmount, uint256 spentAmount)
+    {
+        uint256 buyTokenBalanceBefore = _quote.buyToken.balanceOf(address(this));
+        uint256 sellTokenBalanceBefore = _quote.sellToken.balanceOf(address(this));
+
+
+        (bool success, bytes memory returndata) = swapTarget.call(_quote.swapCallData);
+
+        // Forwarding errors including new custom errors
+        // Taken from: https://ethereum.stackexchange.com/a/111187/73805
+        if (!success) {
+            if (returndata.length == 0) revert();
+            assembly {
+                revert(add(32, returndata), mload(returndata))
+            }
+        }
+
+        boughtAmount = _quote.buyToken.balanceOf(address(this)).sub(buyTokenBalanceBefore);
+        spentAmount = sellTokenBalanceBefore.sub(_quote.sellToken.balanceOf(address(this)));
```

Comment:
Can we calculate the spend amount at the very end of the for loops that use `_fillQuote`? Since we mostly only care about the total amount spent, we can just compare the contracts balance before/after all the quotes have been filled. This will save us a ton of gas since `balanceOf` is an external call which is pretty expensive.

Reply:
Actually we do need to check the balances of each component token separately to avoid the user using "leftover" component amounts in the contract for their issuance / redemption. However for the input/output token one balance call at the end of the loop should indeed be enough.

I will move these balance checks outside of the fillQuote method so that we can reduce the number accordingly (from `2*numComponents` to `numComponents + 1`).

Reply:
I moved the checks of the balance changes outside the fillQuote function and changed it so that the difference of the Input/Output Token balance is only calculated once.

Fell free to check the changes in this commit to see if this is what you had in mind:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/85/commits/af03e21cd9757e9a7fc9d8ca0ca31c1e31206300